### PR TITLE
level compaction expansion

### PR DESCRIPTION
**Pull request by lightmark**
reimplement the compaction expansion on lower level.

Considering such a case:
input level file: 1[B E] 2[F G] 3[H I] 4 [J M]
output level file: 5[A C] 6[D K] 7[L O]

If we initially pick file 2, now we will compact file 2 and 6. But we can safely compact 2, 3 and 6 without expanding the output level.

The previous code is messy and wrong.

In this diff, I first determine the input range [a, b], and output range [c, d],
then we get the range [e,f] = [min(a, c), max(b, d] and put all eligible clean-cut files within [e, f] into this compaction.

**Note: clean-cut means the files don't have the same user key on the boundaries of some files that are not chosen in this compaction**.

### CODE REVIEW

**Comment by siying** [comment]
Can you try to run db_bench timeseries data with small number of series (say 4-8) use a slow benchmark_write_rate_limit to limit the write ratio and populate a multi-level DB and see whether we write much less than before?

**Comment by lightmark** [comment]
```
TEST_TMPDIR=/dev/shm/ ./db_bench --benchmarks=timeseries -target_file_size_base=512000 --key_id_range=4 --write_buffer_size=4000000 --benchmark_write_rate_limit=5000000
grep Expanding /dev/shm//dbbench/LOG | wc -l
```
before 4
after 13

**Comment by siying** [comment]
"level compaction expansion" doesn't tell what the commit is about.

Can you post your benchmark results here?

File: db/version_set.cc [hunk 1]
```diff
@@ -1674,13 +1673,132 @@ void VersionStorageInfo::GetOverlappingInputs(
...
+  }
+
+  const FdWithKeyRange* files = level_files_brief_[level].files;
+  size_t num_files = level_files_brief_[level].num_files;
+  std::vector<size_t> indices(num_files);
+  for (size_t i = 0; i < num_files; i++) {
+    indices[i] = i;
+  }
+  sort(indices.begin(), indices.end(), [&files, user_cmp](size_t i, size_t j) {
+      return user_cmp->Compare(ExtractUserKey(files[i].smallest_key), ExtractUserKey(files[j].smallest_key)) < 0;
```
Comment (siying) [inline db/version_set.cc]: "make format".

File: db/version_set.cc [hunk 2]
```diff
@@ -1674,13 +1673,132 @@ void VersionStorageInfo::GetOverlappingInputs(
...
+    if (begin != nullptr && user_cmp->Compare(file_limit, user_begin) < 0) {
+      // "f" is completely before specified range; skip it
+    } else if (end != nullptr && user_cmp->Compare(file_start, user_end) > 0) {
+      // "f" is completely after specified range; break the loop
+      break;
+    } else {
+      if (overlapping_range_list.empty() ||
+          user_cmp->Compare(file_start, overlapping_range_list.back().second) >
+              0) {
+        overlapping_range_list.push_back(std::pair<Slice,Slice>(file_start, file_limit));
```
Comment (siying) [inline db/version_set.cc]: I think I mentioned this offline. Handling L0 is not needed in this case. This function will be significantly simplified if files won't overlap.

File: db/version_set.cc [hunk 3]
```diff
@@ -1718,17 +1838,29 @@ void VersionStorageInfo::GetOverlappingInputsBinarySearch(
   if (file_index) {
     *file_index = mid;
   }
-  ExtendOverlappingInputs(level, user_begin, user_end, inputs, mid);
+
+  int startIndex, endIndex;
+  if (within_interval) {
+    ExtendFileRangeWithinInterval(level, user_begin, user_end, mid, &startIndex, &endIndex);
```
Comment (siying) [inline db/version_set.cc]: make format"

File: db/version_set.cc [hunk 4]
```diff
@@ -1718,17 +1844,31 @@ void VersionStorageInfo::GetOverlappingInputsBinarySearch(
   if (file_index) {
     *file_index = mid;
   }
-  ExtendOverlappingInputs(level, user_begin, user_end, inputs, mid);
+
+  int startIndex, endIndex;
```
Comment (siying) [inline db/version_set.cc]: naming.

**Pull request closed without merging**